binary_transparency/firmware/cmd/ft_witness/internal/ws/store.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ws contains a Witness Store
+package ws
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/google/trillian-examples/binary_transparency/firmware/api"
+)
+
+const (
+	fileMask = 0755
+)
+
+// Storage is a Witness Storage intended for storing witness checkpoints
+// Currently a simple file is used as a storage mechanism
+type Storage struct {
+	fp        string
+	storeLock sync.Mutex
+}
+
+// NewStorage creates a new WS that uses the given file as DB backend
+// The DB will be initialized if needed.
+func NewStorage(fp string) (*Storage, error) {
+	ws := &Storage{
+		fp: fp,
+	}
+	return ws, ws.init()
+}
+
+// init creates the file storage
+func (ws *Storage) init() error {
+	// Check if the file exists, if not create one
+	f, err := os.OpenFile(ws.fp, os.O_RDWR|os.O_CREATE, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+	defer f.Close()
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+	return nil
+}
+
+// StoreCP saves the given checkpoint into DB.
+func (ws *Storage) StoreCP(wcp api.LogCheckpoint) error {
+
+	ws.storeLock.Lock()
+	defer ws.storeLock.Unlock()
+
+	// Check if file exists, open for write and store the checkpoint
+	f, err := os.OpenFile(ws.fp, os.O_RDWR, fileMask)
+	if err != nil {
+		f.Close()
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+
+	data, err := json.MarshalIndent(wcp, "", " ")
+	if err != nil {
+		return fmt.Errorf("JSON marshaling failed: %w", err)
+	}
+	if _, err := f.Write(data); err != nil {
+		f.Close()
+		return fmt.Errorf("failed to write data to witness db file: %w", err)
+	}
+	if err := f.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+	return nil
+}
+
+// RetrieveCP gets the checkpoint previously stored.
+func (ws *Storage) RetrieveCP() (api.LogCheckpoint, error) {
```

Comment:
This method concerns me; if this is called before `StoreCP` then it will fail, and not as the contract specifies "Must return status code NotFound if no such checkpoint exists.". Please make sure that a new witness which cannot read the log does not explode if get-checkpoint endpoint is called.

Reply:
If there is no Checkpoint in the FT Log at the moment does it return Error ? I suppose it would return Empty Checkpoint right? Please correct me if I am wrong??

I am not sure what you mean by your last statement ?? Can you please clarify ??

Reply:
The behaviour of the log is not relevant to the point I'm making about this `Storage` object. This class as implemented will not live up to its contract if `RetrieveCP` is called before `StoreCP`. The contract says it must return `NotFound` but there is no code here to handle that case.

```
s := NewStorage(createRandomFilePath())
cp, err := s.RetrieveCP()
```

That's an empty/non-existent file path. I would expect cp to be nil, and err to contain `NotFound` given the contract on the interface you defined. As implemented it will be a generic `fmt.Errorf`.

Reply:
Hi Martin,

Let's discuss this on call tomorrow please. Perhaps we should clarify the "Interface Contract" that if No Checkpoint exists in the store, it will return an Empty CheckPoint with no Error.

 Otherwise, the first time the Poll starts it will always return an Empty store and should continue, rather than bailing out!

As mentioned above happy to discuss on the call.

Thanks,
Yogesh

Reply:
As agreed, golden checkpoint retrieved once in the memory at Witness server creation time, Added Mutex for access where two threads can access the gcp and witness store simultaneously!